lib/spack/spack/util/windows_registry.py

```diff
@@ -257,12 +281,42 @@ def find_subkey(self, subkey_name, recursive=True):
         Return:
             the desired subkey as a RegistryKey object, or none
         """
+        # If subkey we're looking for is a regex, perform a regex search
+        try:
+            re.compile(subkey_name)
+            regex = True
+        except re.error:
+            regex = False
 
         if not recursive:
-            return self.get_subkey(subkey_name)
+            ret = self.get_matching_subkeys(subkey_name) if regex else self.get_subkey(subkey_name)
+            return ret[0] if ret else None
+        else:
+            condition = (
+                lambda x: re.match(subkey_name, x.name)
+                if regex
+                else lambda x: x.name == subkey_name
+            )
+            return self._traverse_subkeys(condition)
+
+    def find_subkeys(self, subkey_name, recursive=True):
+        """Exactly the same as find_subkey, except this function tries to match
+        a regex to multiple keys
 
+        Args:
+            subkey_name (str): string representing subkey to be searched for
+            recursive (bool): optional argument, if True, subkey need not be a direct
+                                sub key of this registry entry, and this method will
+                                search all subkeys recursively.
+                                Default is True
+        Return:
+            the desired subkeys as a list of RegistryKey object, or none
+        """
+        if not recursive:
+            return self.get_matching_subkeys(subkey_name)
         else:
-            return self._traverse_subkeys(lambda x: x.name == subkey_name)
+            condition = lambda x: re.match(subkey_name, x.name)
+            return self._traverse_subkeys(condition, collect_found=True)
```

Comment:
This was a deliberate choice on my part. IMO this interface structure is clearer to the caller what the expectations are regarding the search behavior and the nature of the returned object (a full search returning a collection vs a search until the condition is met once and returning single key) based on whether the method is plural or not. In order to reuse `find_subkey` here, we'd need to allow for `find_subkey` to understand the `collect_found` argument and propagate it to `_traverse_subkeys` which doesn't make a lot of sense given the method is only going to return a single subkey.  All of these find methods could ostensibly be a single method with a much more involved interface but I find it simpler to have multiple methods, each of which exposes a specific sliver of functionality. Additionally  I think the code duplication here is pretty minimal, and reusing the single key search method would complicate more than it simplifies. Unless I'm completely off your line of reasoning?

Reply:
Accidentally left the comment below as a review instead of a response, re-commenting here to preserve the thread.

This was a deliberate choice on my part. IMO this interface structure is clearer to the caller what the expectations are regarding the search behavior and the nature of the returned object (a full search returning a collection vs a search until the condition is met once and returning single key) based on whether the method is plural or not. And given the above comment about the dynamic regex determination, I think this is consistent with my response there. In order to reuse find_subkey here, we'd need to allow for find_subkey to understand the collect_found argument and propagate it to _traverse_subkeys which doesn't make a lot of sense given the method is only ever going to return a single subkey. All of these find methods could ostensibly be a single method with a much more involved interface but I find it simpler to have multiple methods, each of which exposes a specific sliver of functionality. Additionally I think the code duplication here is pretty minimal, and reusing the single key search method would complicate more than it simplifies. Unless I'm completely off your line of reasoning?